### PR TITLE
Message Log Widget changes

### PR DESCRIPTION
**Pull request by backendr**
## What will change with this Pull Request?
- removed unused imports
- alphabetised methods
- refactoring
	- code reads down in list format, easier to read, less likely to get lost, tidy
           our eyes are more comfortable when reading tall and narrow columns of text
- removed pass by ref, `QString` uses implicit sharing / C-O-W
- removed logging when moving card from ~`grave` -> `grave`, and~ `rfg` -> `rfg`
- removed join game logging
- construct on first use idiom for const zone names
- conceded/connection lost/connection restored now bold

screen shot 2017-05-09 at 13 41 12

### CODE REVIEW

**Comment by tooomm** [comment]
>removed logging when moving card from `grave` -> `grave`, and `rfg` -> `rfg`

👍 

>removed join game logging

The game# will still show on start of replays?

>conceded/connection lost/connection restored now bold

Do you have a picture handy for the new bold messages?

**Comment by Daenyth** [comment]
The graveyard is an ordered zone, you need to log when people change the
ordering of cards.

**Comment by tooomm** [comment]
True for legacy formats... http://mtg.gamepedia.com/Graveyard#Order_of_cards_in_the_graveyard

**Comment by backendr** [comment]
@tooomm forgot to add one. Added to initial message.

**Comment by backendr** [comment]
@Daenyth added graveyard logging back

File: cockatrice/src/messagelogwidget.cpp [hunk 1]
```diff
@@ -3,190 +3,87 @@
...
+        int position,
+        bool ownerChange) const
 {
     bool cardNameContainsStartZone = false;
     QString fromStr;
-    QString startName = zone->getName();
+    QString zoneName = zone->getName();
 
-    if (startName == "table")
+    if (zoneName == tableConstant())
```
Comment (ZeldaZach) [inline cockatrice/src/messagelogwidget.cpp]: Instead of declaring `tableConstant()` and the other constants as functions, define them as full cap constants at the top of the file maybe?
```
#define TABLE "table"
#define GRAVEYARD "grave"
etc
```

File: cockatrice/src/messagelogwidget.cpp [hunk 2]
```diff
@@ -219,278 +116,379 @@ QPair<QString, QString> MessageLogWidget::getFromStr(CardZone *zone, QString car
...
+
+void MessageLogWidget::logDoMoveCard(LogMoveCard &lmc)
+{
+    QString startZone = lmc.startZone->getName();
+    QString targetZone = lmc.targetZone->getName();
+    bool ownerChanged = lmc.startZone->getPlayer() != lmc.targetZone->getPlayer();
+
+    // do not log if moved within the same zone
+    if ((startZone == tableConstant() && targetZone == tableConstant() && !ownerChanged)
+            || (startZone == handConstant() && targetZone == handConstant())
```
Comment (ZeldaZach) [inline cockatrice/src/messagelogwidget.cpp]: owner changed still might be a thing here, as I _could_ drag a card from my hand to my opponents hand (no idea why, but some games might do that?)

File: cockatrice/src/messagelogwidget.cpp (hunk 1 above)
Comment (backendr) [inline cockatrice/src/messagelogwidget.cpp]: http://stackoverflow.com/questions/4103086/defining-constants-in-c

File: cockatrice/src/messagelogwidget.cpp [hunk 3]
```diff
@@ -508,139 +506,357 @@ void MessageLogWidget::logRevealCards(Player *player, CardZone *zone, int cardId
...
-                    str = tr("%1 reveals %2 to %3.");
-            appendHtmlServerMessage(str.arg(sanitizeHtml(player->getName())).arg(zone->getTranslatedName(true, CaseRevealZone)).arg(sanitizeHtml(otherPlayer->getName())));
-        } else {
-            appendHtmlServerMessage((
-              tr("%1 reveals %2.")
-            ).arg(sanitizeHtml(player->getName()))
-             .arg(zone->getTranslatedName(true, CaseRevealZone)));
-        }
+        if (otherPlayer)
+            appendHtmlServerMessage(QString(tr("%1 reveals %2 to %3."))
```
Comment (ZeldaZach) [inline cockatrice/src/messagelogwidget.cpp]: TR will return a QString itself, no need to wrap it (https://doc.qt.io/qt-5/qobject.html#tr)

File: cockatrice/src/messagelogwidget.cpp (hunk 2 above)
Comment (backendr) [inline cockatrice/src/messagelogwidget.cpp]: This is not currently possible in the master branch.

File: cockatrice/src/messagelogwidget.cpp [hunk 4]
```diff
@@ -508,139 +506,357 @@ void MessageLogWidget::logRevealCards(Player *player, CardZone *zone, int cardId
...
-    connect(player, SIGNAL(logAttachCard(Player *, QString, Player *, QString)), this, SLOT(logAttachCard(Player *, QString, Player *, QString)));
-    connect(player, SIGNAL(logUnattachCard(Player *, QString)), this, SLOT(logUnattachCard(Player *, QString)));
-    connect(player, SIGNAL(logDumpZone(Player *, CardZone *, int)), this, SLOT(logDumpZone(Player *, CardZone *, int)));
-    connect(player, SIGNAL(logStopDumpZone(Player *, CardZone *)), this, SLOT(logStopDumpZone(Player *, CardZone *)));
-    connect(player, SIGNAL(logDrawCards(Player *, int)), this, SLOT(logDrawCards(Player *, int)));
-    connect(player, SIGNAL(logUndoDraw(Player *, QString)), this, SLOT(logUndoDraw(Player *, QString)));
-    connect(player, SIGNAL(logRevealCards(Player *, CardZone *, int, QString, Player *, bool)), this, SLOT(logRevealCards(Player *, CardZone *, int, QString, Player *, bool)));
-    connect(player, SIGNAL(logAlwaysRevealTopCard(Player *, CardZone *, bool)), this, SLOT(logAlwaysRevealTopCard(Player *, CardZone *, bool)));
+    connect(player, SIGNAL(logSay(Player *, QString)),
```
Comment (ZeldaZach) [inline cockatrice/src/messagelogwidget.cpp]: I kind of liked how the connects were all in line, feels like waste of space now

File: cockatrice/src/messagelogwidget.cpp (hunk 1 above)
Comment (ZeldaZach) [inline cockatrice/src/messagelogwidget.cpp]: In the header file we can put them as constant public variables. I've never seen it done as a function like this before and it just feels a bit off

File: cockatrice/src/messagelogwidget.cpp (hunk 2 above)
Comment (ZeldaZach) [inline cockatrice/src/messagelogwidget.cpp]: Darn, a feature to be done in the future

File: cockatrice/src/messagelogwidget.cpp (hunk 1 above)
Comment (backendr) [inline cockatrice/src/messagelogwidget.cpp]: This is the construct on first use idiom.
http://stackoverflow.com/questions/335369/finding-c-static-initialization-order-problems

I dont feel adding "global" consts here is a good idea as you suggest. Instead it might be better to start a `constants.h/cpp` that can be accessed from anywhere.

**Comment by tooomm** [comment]
Can you please combine your code comments in one review @ZeldaZach ?
You can also just use the review to comment, in case you don't want to request changes or approve things.

I just got hammered with mails :)

File: cockatrice/src/messagelogwidget.cpp (hunk 1 above)
Comment (ZeldaZach) [inline cockatrice/src/messagelogwidget.cpp]: I'd be fine with a global constants file somewhere

Comment (backendr) [inline cockatrice/src/messagelogwidget.cpp]: OK. That is not part of this PR though.

File: cockatrice/src/messagelogwidget.cpp [hunk 5]
```diff
@@ -508,139 +506,357 @@ void MessageLogWidget::logRevealCards(Player *player, CardZone *zone, int cardId
...
+    appendHtmlServerMessage(str.arg(sanitizeHtml(player->getName()))
+                                    .arg(cardLink(card->getName())));
+}
+
+void MessageLogWidget::logSetPT(Player *player, CardItem *card, QString newPT)
+{
+    if (currentContext == MessageContext_MoveCard)
+        moveCardPT.insert(card, newPT);
+    else
+        appendHtmlServerMessage(QString(tr("%1 sets PT of %2 to %3."))
```
Comment (ctrlaltca) [inline cockatrice/src/messagelogwidget.cpp]: QString() around tr() can be removed.

**Comment by tooomm** [comment]
Since we have now highights for several (more) important messages...
I feel like we missed `deck load` and should also print them in bold.
